File: rust/src/methods/notifications.rs
//! Notification methods (OFS-6000).

use crate::client::{Client, IdParams, WalletParams, encode_peer_id};
use crate::error::Result;
use openfiat_crypto::Keypair;
use openfiat_notifications::events::{
    DeliveryReport, SignedDeliveryReport, SignedSubscriptionUpdate, SubscriptionUpdate,
};
use openfiat_notifications::{DeliveryReceipt, Subscription};
use openfiat_types::PeerId;

impl Client {
    pub async fn get_subscription(&self, wallet: &PeerId) -> Result<Option<Subscription>> {
        self.call(
            "getSubscription",
            WalletParams {
                wallet: encode_peer_id(wallet),
            },
        )
        .await
    }

    pub async fn get_delivery_receipts_by_wallet(
        &self,
        wallet: &PeerId,
    ) -> Result<Vec<DeliveryReceipt>> {
        self.call(
            "getDeliveryReceiptsByWallet",
            WalletParams {
                wallet: encode_peer_id(wallet),
            },
        )
        .await
    }

    pub async fn get_delivery_receipt(
        &self,
        id: impl Into<String>,
    ) -> Result<Option<DeliveryReceipt>> {
        self.call("getDeliveryReceipt", IdParams { id: id.into() })
            .await
    }

    pub async fn send_subscription_update(
        &self,
        update: SubscriptionUpdate,
        keypair: &Keypair,
    ) -> Result<()> {
        let signed = SignedSubscriptionUpdate::sign(update, keypair);
        self.send_signed("sendSubscriptionUpdate", &signed).await
    }

    pub async fn send_delivery_report(
        &self,
        report: DeliveryReport,
        keypair: &Keypair,
    ) -> Result<()> {
        let signed = SignedDeliveryReport::sign(report, keypair);
        self.send_signed("sendDeliveryReport", &signed).await
    }
}

File: typescript/src/methods/notifications.ts
import type { Client } from "../client.js";
import { type Keypair, sign } from "../crypto.js";
import {
  toBase58,
  type DeliveryReceipt,
  type DeliveryReport,
  type SignedDeliveryReport,
  type SignedSubscriptionUpdate,
  type Subscription,
  type SubscriptionUpdate,
} from "../types.js";

/** Base64-encodes a peer ID the same way `WalletParams` expects it on the wire. */
function encodePeerId(peerId: Uint8Array): string {
  return Buffer.from(peerId).toString("base64");
}

export async function getSubscription(client: Client, wallet: Uint8Array): Promise<Subscription | null> {
  return client.call("getSubscription", { wallet: encodePeerId(wallet) });
}

export async function getDeliveryReceipt(client: Client, id: string): Promise<DeliveryReceipt | null> {
  return client.call("getDeliveryReceipt", { id });
}

export async function getDeliveryReceiptsByWallet(
  client: Client,
  wallet: Uint8Array,
): Promise<DeliveryReceipt[]> {
  return client.call("getDeliveryReceiptsByWallet", { wallet: encodePeerId(wallet) });
}

/** §11: publish this wallet's own notification subscription preferences. */
export async function sendSubscriptionUpdate(
  client: Client,
  update: SubscriptionUpdate,
  keypair: Keypair,
): Promise<void> {
  const bytes = new TextEncoder().encode(JSON.stringify(update));
  const signature = await sign(keypair, bytes);
  const signed: SignedSubscriptionUpdate = { update, signature: toBase58(signature) };
  return client.sendSigned("sendSubscriptionUpdate", signed);
}

/**
 * A provider reports the outcome of one delivery attempt — `keypair`
 * must already be registered as a Notification Provider (see
 * {@link providers.sendProviderRegister}) or the node will reject it.
 */
export async function sendDeliveryReport(
  client: Client,
  report: DeliveryReport,
  keypair: Keypair,
): Promise<void> {
  const bytes = new TextEncoder().encode(JSON.stringify(report));
  const signature = await sign(keypair, bytes);
  const signed: SignedDeliveryReport = { report, signature: toBase58(signature) };
  return client.sendSigned("sendDeliveryReport", signed);
}
